src/reconcile/rec_visibility.c

```diff
@@ -281,6 +282,13 @@ __wt_rec_upd_select(WT_SESSION_IMPL *session, WT_RECONCILE *r, WT_INSERT *ins, v
         if (WT_TXNID_LT(max_txn, txnid))
             max_txn = txnid;
 
+        /*
+         * Special handling for application threads evicting their own updates.
+         */
+        if (!is_hs_page && F_ISSET(r, WT_REC_APP_UPDATED_SNAPSHOT) && txnid == session_txnid) {
+            has_newer_updates = true;
```

Comment:
Would something similar to this be enough without copying out and allocating a new snapshot for each reconciliation?

i.e: the correctness problem with doing snapshot based reconciliation using an application thread is that it's possible to evict uncommitted content associated with the currently running transaction. Could we just use the existing snapshot and keep this check?

Reply:
@agorrod - We can just add this check and force isolation level to improve the efficacy of eviction by application threads. However, it will not match the performance we can achieve by taking an updated snapshot. Essentially, we will not be able to evict committed changes belonging to transactions in snapshot or transaction ids higher than `snap_max`.

Reply:
I think that compromise is acceptable given the reduced complexity of the solution.
If we encounter cases where it's causing problems we can do something more similar to the complete solution here.

Reply:
@agorrod I have pushed 38dc1cb where we are use existing snapshot for application threads.